427 - Rename SearchManagerContext state to searchParams and simplify useSearchManager implementation

nozomione opened the pull request:
## Issue Number

Closes #427

Base branch: PR #430

## Purpose/Implementation Notes

Changes include:
- Updated `SearchManagerContext`:
  - Renamed the `search` state to `searchParams`
  - Deleted the `filterOrders` state 
- Optimized the `useSearchManager` hook:
  - Replaced the `updateSearchQuery` method with a new `useSyncSearch` hook, which syncs the latest search parameters with the search results page's URL 
  - Replaced the `filterOrders` state with a local variable `filterOrders` in the `toggleFilter` method and deleted `addFilter`, `removeFilter`, `updateFilterOrders` methods that were used for this replaced state
  - Adjusted `helpers/getSearchQueryForAPI` to include the parameter values when defined
- Decouple the experiment page's `ref` query (used to indicate it's from the search results) from the `useSearchManager` hook in the following components:
  - `SearchCardHeader`
  - `SearchCardFooter`
- Adjusted the `pages/search` accordingly 
## Types of changes

- Refactor (addresses code organization and design mentioned in corresponding issue)

## Functional tests

N/A

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

N/A

nozomione commented on src/components/shared/SearchCard/SearchCardFooter.js:
```diff
@@ -1,19 +1,19 @@
-    push(`/experiments/${accessionCode}/${formatURLString(title)}`)
+    push({
+      pathname: `/experiments/${accessionCode}/${formatURLString(title)}`,
+      query: { ref: 'view-samples' }
```
Removed the use of `useSearchManager` for the search page reference and adjusted `gtag.trackLink` accordingly.

nozomione commented on src/components/shared/SearchCard/SearchCardHeader.js:
```diff
@@ -35,7 +32,12 @@ export const SearchCardHeader = ({ experiment, isLinked = false }) => {
         {isLinked ? (
           <Anchor
-            href={`experiments/${accessionCode}/${formatURLString(title)}`}
+            href={{
+              pathname: `experiments/${accessionCode}/${formatURLString(
+                title
+              )}`,
+              query: { ref: 'search' }
+            }}
```
Removed the use of `useSearchManager` for the search page reference and adjusted `gtag.trackLink` accordingly.

nozomione commented:
I've applied your feedback and this PR is ready for another look. Thank you, David!

nozomione commented on src/components/SearchResults/SearchFilterList/HideNonDownloadableExperiments.js:
```diff
@@ -0,0 +1,28 @@
+import { CheckBox } from 'components/shared/CheckBox'
+import { options } from 'config'
+
+export const HideNonDownloadableExperiments = () => {
```
**NOTE:** I've updated this component name and adjusted its import.

nozomione commented on src/hooks/useSearchManager.js:
```diff
@@ -1,155 +1,150 @@
-    setSearch({ ...search })
-    updateSearchQuery()
+  // update a single parameter value
+  const updateParamValue = (paramName, newValue) => {
+    setSearchParams({
+      ...searchParams,
+      [paramName]: newValue
+    })
```
I've defined a separate method here for the `num_downloadable_samples__gt` key, as it's no longer part of the facet filter items, and removed its filtering logic from `toggleFilter` accordingly. The method is generic and simply updates the value of a single search parameter to the given value. 

Let me know if you have any suggestions for a better name for this method!

davidsmejia replied:
`updateSearchParam` would work, since we are updating an arbitrary param.

nozomione commented:
I've applied the 1:1 feedback and this PR is ready for another look. Thank you, David!

davidsmejia reviewed (commented):
- can you remove the concept of `isBooleanValue` and create a setFilter or setFilterValue instead and call this from your SearchBooleanFilter, this would update the offset but not the filter_order
- you can further simplify the toggleFilter by defining helpers for merging / removing param values ex: `getQueryParamValueWith` and `getQueryParamValueWithout` this should support default values that are not arrays

nozomione commented:
I've included the following changes: 
- Added `helpers/getQueryParamValue.js` with `getQueryParamValueWith` and `getQueryParamValueWithout` for handling array checks on filter addition and deletion and adjusted `toggleFilter` to use these helpers
- Added  a new `updateFilterValue` method and used it in `SearchBooleanFilter`, and removed the boolean filter logic from `toggleFilter`

This PR is ready for another look. Thank you, David!

davidsmejia commented on src/helpers/getQueryParamValue.js:
```diff
@@ -0,0 +1,26 @@
+// - the array becomes empty after removal
+// - 'prev' is equal to 'toRemove'
+// Otherwise, return 'prev' unchanged
+export const getQueryParamValueWithout = (prev, toRemove) => {
+  if (prev === toRemove) return undefined
+
+  if (Array.isArray(prev)) {
+    const updatedPrev = prev.filter((v) => v !== toRemove)
+    return updatedPrev.length === 0 ? undefined : updatedPrev
+  }
+
+  return prev
+}
```
If the returning array is length 1 after removing we should just return the non-array value.

davidsmejia reviewed (approved):
LGTM just need that one small change

nozomione commented:
I've included the change and am now merging this PR into `dev`. Thank you!
